zebrad/src/components/mempool/downloads.rs

```diff
@@ -199,6 +199,28 @@ where
     /// Returns the action taken in response to the queue request.
     #[instrument(skip(self, txid), fields(txid = %txid))]
     pub fn download_and_verify(&mut self, txid: UnminedTxId) -> DownloadAction {
+        self.download_if_needed_and_verify(txid, None)
+    }
+
+    /// Queue a transaction for verification.
+    ///
+    /// Returns the action taken in response to the queue request.
+    #[instrument(skip(self, tx))]
+    pub fn verify(&mut self, tx: UnminedTx) -> DownloadAction {
+        self.download_if_needed_and_verify(tx.id, Some(tx))
+    }
+
+    /// Queue a transaction for download, if needed, and verification.
+    ///
+    /// If `tx` is None then the transaction will be download from the given `txid`
+    /// and verified. Otherwise, the given transaction is verified.
+    ///
+    /// Returns the action taken in response to the queue request.
+    fn download_if_needed_and_verify(
+        &mut self,
+        txid: UnminedTxId,
+        tx: Option<UnminedTx>,
```

Comment:
Optional style: Sometimes we have a duplicate transaction ID here - `txid` and `tx.id`.

We could model the request as an enum instead.
Then we could re-use the enum in the service interface as part of #2728.

```rust
enum GossipedTx {
    Id(UnminedTxId),
    Tx(UnminedTx),
}

impl GossipedTx {
    fn id(&self) -> UnminedTxId { ... }
}

impl From<UnminedTxId> for GossipedTx  { ... }
impl From<UnminedTx> for GossipedTx { ... }
```

Reply:
Great idea! Added in 239820dc.

The only thing I'm not sure is what method to expose now. Before this it had `download_and_verify` and `verify`, now there's just `download_if_needed_and_verify` which does describe what it does but it's a bit ugly. Suggestions are welcome.

Reply:
@dconnolly Can you help @conradoplg out here now that @teor2345 is out?

Reply:
This can be addressed in https://github.com/ZcashFoundation/zebra/pull/2741 since it refactors this part of the code